x/wasm/internal/keeper/submsg_test.go

```diff
@@ -0,0 +1,375 @@
+
+	deposit := sdk.NewCoins(sdk.NewInt64Coin("denom", 100000))
+	contractStart := sdk.NewCoins(sdk.NewInt64Coin("denom", 40000))
+
+	creator := createFakeFundedAccount(t, ctx, accKeeper, bankKeeper, deposit)
+	creatorBalance := deposit.Sub(contractStart)
+	_, _, fred := keyPubAddr()
+
+	// upload code
+	reflectCode, err := ioutil.ReadFile("./testdata/reflect.wasm")
+	require.NoError(t, err)
+	codeID, err := keeper.Create(ctx, creator, reflectCode, "", "", nil)
+	require.NoError(t, err)
+	require.Equal(t, uint64(1), codeID)
+
+	// creator instantiates a contract and gives it tokens
+	contractAddr, _, err := keeper.Instantiate(ctx, codeID, creator, nil, []byte("{}"), "reflect contract 1", contractStart)
+	require.NoError(t, err)
+	require.NotEmpty(t, contractAddr)
+
+	// check some account values
+	checkAccount(t, ctx, accKeeper, bankKeeper, contractAddr, contractStart)
+	checkAccount(t, ctx, accKeeper, bankKeeper, creator, creatorBalance)
+	checkAccount(t, ctx, accKeeper, bankKeeper, fred, nil)
+
+	// creator can send contract's tokens to fred (using SendMsg)
+	msg := wasmvmtypes.CosmosMsg{
+		Bank: &wasmvmtypes.BankMsg{
+			Send: &wasmvmtypes.SendMsg{
+				ToAddress: fred.String(),
+				Amount: []wasmvmtypes.Coin{{
+					Denom:  "denom",
+					Amount: "15000",
+				}},
+			},
+		},
+	}
+	reflectSend := ReflectHandleMsg{
+		ReflectSubCall: &reflectSubPayload{
+			Msgs: []wasmvmtypes.SubMsg{{
+				ID:  7,
+				Msg: msg,
+			}},
+		},
+	}
+	reflectSendBz, err := json.Marshal(reflectSend)
+	require.NoError(t, err)
+	_, err = keeper.Execute(ctx, contractAddr, creator, reflectSendBz, nil)
+	require.NoError(t, err)
+
+	// fred got coins
+	checkAccount(t, ctx, accKeeper, bankKeeper, fred, sdk.NewCoins(sdk.NewInt64Coin("denom", 15000)))
+	// contract lost them
+	checkAccount(t, ctx, accKeeper, bankKeeper, contractAddr, sdk.NewCoins(sdk.NewInt64Coin("denom", 25000)))
+	checkAccount(t, ctx, accKeeper, bankKeeper, creator, creatorBalance)
+
+	// TODO: query the reflect state to ensure the result was stored
+	query := ReflectQueryMsg{
+		SubCallResult: &SubCall{ID: 7},
+	}
+	queryBz, err := json.Marshal(query)
+	require.NoError(t, err)
+	queryRes, err := keeper.QuerySmart(ctx, contractAddr, queryBz)
+	require.NoError(t, err)
+
+	var res wasmvmtypes.Reply
+	err = json.Unmarshal(queryRes, &res)
+	require.NoError(t, err)
+	assert.Equal(t, uint64(7), res.ID)
+	assert.Empty(t, res.Result.Err)
+	require.NotNil(t, res.Result.Ok)
+	sub := res.Result.Ok
+	assert.Empty(t, sub.Data)
+	require.Len(t, sub.Events, 3)
+
+	transfer := sub.Events[0]
+	assert.Equal(t, "transfer", transfer.Type)
+	assert.Equal(t, wasmvmtypes.EventAttribute{
+		Key:   "recipient",
+		Value: fred.String(),
+	}, transfer.Attributes[0])
+
+	sender := sub.Events[1]
+	assert.Equal(t, "message", sender.Type)
+	assert.Equal(t, wasmvmtypes.EventAttribute{
+		Key:   "sender",
+		Value: contractAddr.String(),
+	}, sender.Attributes[0])
+
+	// where does this come from?
+	module := sub.Events[2]
```

Comment:
The event-filter is built into the [handler](https://github.com/CosmWasm/wasmd/blob/master/x/wasm/handler.go#L51) currently. It guess we need it in the keeper now.

Reply:
We don't really need it there, I guess it would just be cleaner?

Let's make another issue for event filtering. Or just investigate more during https://github.com/CosmWasm/wasmd/issues/445

Reply:
I will add a new issue for this.